atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java

```diff
@@ -132,18 +136,24 @@ public void close() {}
 
             // RangeMap needs a little help with weird 1-node, 1-vnode, this-entire-feature-is-useless case
             if (tokenRanges.size() == 1) {
-                String onlyEndpoint = Iterables.getOnlyElement(
-                        Iterables.getOnlyElement(tokenRanges).getEndpoints());
-                InetSocketAddress onlyHost = getAddressForHost(onlyEndpoint);
+                EndpointDetails onlyEndpoint = Iterables.getOnlyElement(
+                        Iterables.getOnlyElement(tokenRanges).getEndpoint_details());
+                InetSocketAddress onlyHost = getAddressForHost(onlyEndpoint.getHost());
                 newTokenRing.put(Range.all(), ImmutableList.of(onlyHost));
                 servers.add(onlyHost);
+                hostToDatacentersThisRefresh.put(onlyHost, onlyEndpoint.getDatacenter());
             } else { // normal case, large cluster with many vnodes
                 for (TokenRange tokenRange : tokenRanges) {
-                    List<InetSocketAddress> hosts = tokenRange.getEndpoints().stream()
-                            .map(this::getAddressForHostThrowUnchecked)
-                            .collect(Collectors.toList());
-
+                    Map<InetSocketAddress, String> hostToDatacentersOnThisTokenRange = KeyedStream.of(
+                                    tokenRange.getEndpoint_details())
+                            .mapKeys(EndpointDetails::getHost)
+                            .mapKeys(this::getAddressForHostThrowUnchecked)
+                            .map(EndpointDetails::getDatacenter)
+                            .collectToMap();
```

Comment:
please trace log this so we can more easily reason about it if needed

Reply:
discussed offline: moved to once _per iteration_ as opposed to once _per interval_, but added!